### PR TITLE
ARROW-7268: [Rust] Add `custom_metadata` field from IPC message to Schema.

### PR DESCRIPTION
**Pull request by grundprinzip**
This patch propagates the `custom_metadata` field from the IPC message
to the Schema type. It adds the necessary serailization and
deserialization logic and tests.

### CODE REVIEW

File: rust/arrow/src/datatypes.rs [hunk 1]
```diff
@@ -949,28 +984,59 @@ impl Schema {
...
+                Ok(Self { fields, metadata })
             }
             _ => Err(ArrowError::ParseError(
                 "Invalid json value type for schema".to_string(),
             )),
         }
     }
+
+    /// Parse a `metadata` definition from a JSON representation
+    fn _from_metadata(json: &Value) -> Result<HashMap<String, String>> {
```
Comment (nevi-me) [inline rust/arrow/src/datatypes.rs]: We don't prefix functions with `_` if they're private. I'm assuming this is why you did this?

```suggestion
    fn from_metadata(json: &Value) -> Result<HashMap<String, String>> {
```

**Review by nevi-me — COMMENTED** [review]
Hi @grundprinzip, one minor comment on a private function name

File: rust/arrow/src/datatypes.rs (hunk 1 above)
Comment (grundprinzip) [reply rust/arrow/src/datatypes.rs]: Done

**Review by paddyhoran — APPROVED** [review]
LGTM, thanks @grundprinzip

**Review by nevi-me — APPROVED** [review]
LGTM

**Pull request closed without merging**